pkg/function/service.go

```diff
@@ -20,9 +45,26 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*functionpb.ReadyRe
 	return &functionpb.ReadyResponse{Ready: true}, nil
 }
 
-// DoFn applies a function to each datum element
-func (fs *Service) DoFn(ctx context.Context, d *functionpb.Datum) (*functionpb.DatumList, error) {
-	messages, err := fs.Mapper.HandleDo(ctx, d.GetKey(), d.GetValue())
+// MapFn applies a function to each datum element
+func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.DatumList, error) {
+	var key string
+
+	// get key from gPRC metadata
+	if grpcMD, ok := metadata.FromIncomingContext(ctx); ok {
+		keyValue := grpcMD.Get(DatumKey)
+		if len(keyValue) > 1 {
+			return nil, fmt.Errorf("expect one key but got multiple keys")
```

Comment:
let's do a `%d` on len of the keys and perhaps even list it via `%v` ? will make debugging easier.